### PR TITLE
feat: enable glm5 router gemm

### PR DESCRIPTION
**Pull request by b8zhong**
## 📌 Description

It uses hidden_dim = 6144, need to add it to allowlist (ideally could have router gemm API to avoid duplication), but this is faster for now.

On the performance:

###  GLM-5 Router GEMM Benchmark (Flashinfer vs Torch) with CUPTI benchmark

| M (Tokens) | N (Experts) | K (Dim) | Torch (ms) | Flashinfer (ms) | Winner | Speedup |
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
| 1 | 256 | 6144 | 6.048 | 2.976 | Flashinfer | +103.2% |
| 2 | 256 | 6144 | 6.272 | 3.137 | Flashinfer | +99.9% |
| 3 | 256 | 6144 | 6.272 | 3.584 | Flashinfer | +75.0% |
| 4 | 256 | 6144 | 6.304 | 3.872 | Flashinfer | +62.8% |
| 5 | 256 | 6144 | 6.304 | 4.032 | Flashinfer | +56.4% |
| 6 | 256 | 6144 | 6.304 | 4.480 | Flashinfer | +40.7% |
| 7 | 256 | 6144 | 6.304 | 4.480 | Flashinfer | +40.7% |
| 8 | 256 | 6144 | 6.304 | 4.832 | Flashinfer | +30.5% |
| 9 | 256 | 6144 | 7.040 | 5.216 | Flashinfer | +35.0% |
| 10 | 256 | 6144 | 7.040 | 5.408 | Flashinfer | +30.2% |
| 11 | 256 | 6144 | 7.040 | 7.616 | torch | -7.6% |
| 12 | 256 | 6144 | 7.040 | 7.936 | torch | -11.3% |
| 13 | 256 | 6144 | 7.040 | 8.128 | torch | -13.4% |
| 14 | 256 | 6144 | 7.040 | 8.288 | torch | -15.1% |
| 15 | 256 | 6144 | 7.072 | 8.544 | torch | -17.2% |
| 16 | 256 | 6144 | 6.400 | 8.768 | torch | -27.0% |

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

Add new hidden dim

```
pytest tests/model_optimizations/test_dsv3_router_gemm.py
============================================================================ test session starts =============================================================================
platform linux -- Python 3.12.3, pytest-9.0.3, pluggy-1.6.0
rootdir: /sgl-workspace/flashinfer
configfile: pytest.ini
plugins: anyio-4.13.0, typeguard-4.5.1
collected 57 items                                                                                                                                                           

tests/model_optimizations/test_dsv3_router_gemm.py .........................................................                                                           [100%]

============================================================================= 57 passed in 2.44s =============================================================================
```

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

## Summary by CodeRabbit

* **New Features**
  * Added support for a new GLM-MoE-DSA router GEMM configuration (6144×256) and its trace/reference implementation.

* **Improvements**
  * Generalized router GEMM shape validation and added new dispatch/export pathway for the 6144×256 variant.
  * Changed defaults for existing router GEMM launch behavior to enable optimized dispatch.
  * Replaced low-level GPU sync primitives with updated synchronization calls for CUDA 9.0+. 

* **Tests**
  * Expanded unit tests to cover the new 6144×256 kernel and additional negative validation cases.

[Review Change Stack](https://app.coderabbit.ai/change-stack/flashinfer-ai/flashinfer/pull/3185)

### CODE REVIEW

File: flashinfer/gemm/routergemm.py [hunk 1]
```diff
@@ -119,13 +140,26 @@ def mm_M1_16_K7168_N256(
         mat_a: torch.Tensor,
         mat_b: torch.Tensor,
         out: torch.Tensor,
-        launch_with_pdl: bool = False,
+        launch_with_pdl: bool = True,
```
[the remark this answers is not in the document]
Comment (b8zhong) [reply flashinfer/gemm/routergemm.py:144]: Um, I think many of the GEMM APIs use PDL on by default? So I think it's ok...

File: tests/model_optimizations/test_dsv3_router_gemm.py [hunk 2]
```diff
@@ -1,20 +1,21 @@
 import torch
 import pytest
 from flashinfer.dsv3_ops import mm_M1_16_K7168_N128, mm_M1_16_K7168_N256
+from flashinfer.gemm import mm_M1_16_K6144_N256
```
[the remark this answers is not in the document]
Comment (b8zhong) [reply tests/model_optimizations/test_dsv3_router_gemm.py]: I think this was for backwards compatibility before (?), can add this if really needed

Comment (leejnau) [inline tests/model_optimizations/test_dsv3_router_gemm.py]: **API namespace consistency:** Since `mm_M1_16_K6144_N256` is a GEMM specialization and is not DeepSeek-specific, I think `flashinfer.gemm` should be the canonical public import path for this router GEMM family. Could we import all three router GEMM variants from `flashinfer.gemm` here instead of mixing `flashinfer.dsv3_ops` for the K7168 variants and `flashinfer.gemm` for the new K6144 variant?

  ```suggestion
  from flashinfer.gemm import (
      mm_M1_16_K6144_N256,
      mm_M1_16_K7168_N128,
      mm_M1_16_K7168_N256,
  )

Comment (sricketts) [reply tests/model_optimizations/test_dsv3_router_gemm.py]: Fair point, but even with that change, we still have the problem that the test file is called `test_dsv3_router_gemm.py`.

I think the original intent behind the `dsv3_ops` namespace was to:
1. signal that these ops are specializations for DSV3
2. group all such specializations in one place

I'm not sure how much value we've gotten from (2).

Maybe what we should do is:
* rename this file to `test_specialized_router_gemms.py` or something like that
* either:
  * add a `gemm.specialized_router_gemm` namespace and move all of these gemms under that; or
  * stop trying to get fancy with namespaces and just use the top level `gemm` namespace

@dhiraj113 @aleozlx do you have any thoughts?

Also cc @nvmbreughe who was involved in the original design

Comment (leejnau) [reply tests/model_optimizations/test_dsv3_router_gemm.py]: Yes, agree on file name.  Maybe `test_specialized_router_gemms.py` or even just `test_router_gemms.py` would be better.  I searched both sglang and vllm and see no uses of `flashinfer.dsv3_ops`.  
My vote would just be `flashinfer.gemm` for simplicity (no sub-namespaces), but `gemm.specialized_router_gemm` seems fine also.

Comment (dhiraj113) [reply tests/model_optimizations/test_dsv3_router_gemm.py]: I would generally prefer the `gemm` namespace, but here the API is implicit about the `output datatype` and is specializing on it for the specific NN model, so model specific namespace also makes sense. The `__init__.py` files exposes both the `gemm` namepsace as well as `dsv3` namespace for the `dsv3` router gemm operation.

Renaming the test file make sense though, as it is testing sizes/datatypes from various models.

In addition, all these APIs should be combined into one I think, with `output datatype` added as an argument to the API.

**Review by dhiraj113 — APPROVED** [review]
LGTM.

**Comment by dhiraj113** [comment]
Approved, but @b8zhong, you mention above that flashinfer is slower than torch for `M -> 11 to 16`,  is this before your change or after?

**Comment by b8zhong** [comment]
@dhiraj113 It's before this change. This PR doesn't change perf code, only gating hidden dim allowlist for this kernel. This original kernel (for DSV3 hidden size) had speedup compared to Torch when num tokens <= 16, now only until 8

File: flashinfer/gemm/routergemm.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (b8zhong) [reply flashinfer/gemm/routergemm.py:144]: Just changed it for all

File: csrc/dsv3_router_gemm.cu [hunk 3]
```diff
@@ -56,15 +56,16 @@ struct LoopUnroller<kEnd, kEnd, kNumExperts, kHiddenDim> {
...
                             bool launch_with_pdl) {
+  // Kernel processes VPT(=8 for bf16) * kBlockSize(=128) = 1024 elements of K per iteration.
+  static_assert(kHiddenDim % 1024 == 0, "kHiddenDim must be a multiple of 1024");
   int const num_tokens = mat_a.sizes()[0];
   int const num_experts = mat_b.sizes()[1];
   int const hidden_dim = mat_a.sizes()[1];
   auto const out_dtype_ = out.dtype();
   auto const data_type = mat_a.dtype();
-  constexpr int kHiddenDim = 7168;
   std::vector<int64_t> output_size = {mat_a.sizes()[0], mat_b.sizes()[1]};
```
[the remark this answers is not in the document]
Comment (b8zhong) [reply csrc/dsv3_router_gemm.cu]: Deleted

File: tests/model_optimizations/test_router_gemms.py [hunk 4]
```diff
@@ -1,20 +1,21 @@
...
 @pytest.mark.parametrize("num_tokens", [1, 2, 3, 5, 8, 13, 16])
 @pytest.mark.parametrize(
-    "num_experts,output_dtype,fn_to_test",
+    "num_experts,output_dtype,hidden_dim,fn_to_test",
     (
-        [256, torch.float32, mm_M1_16_K7168_N256],
-        [128, torch.bfloat16, mm_M1_16_K7168_N128],
+        [256, torch.float32, 7168, mm_M1_16_K7168_N256],
+        [128, torch.bfloat16, 7168, mm_M1_16_K7168_N128],
+        [256, torch.float32, 6144, mm_M1_16_K6144_N256],
```
[the remark this answers is not in the document]
Comment (b8zhong) [reply tests/model_optimizations/test_router_gemms.py:19]: Added negative tests

File: flashinfer/gemm/routergemm.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (b8zhong) [reply flashinfer/gemm/routergemm.py:144]: Changed it for N128

File: flashinfer/gemm/routergemm.py [hunk 5]
```diff
@@ -229,6 +263,54 @@ def mm_M1_16_K7168_N256(
     )
 
 
+@backend_requirement({}, common_check=_mm_M1_16_K6144_N256_shape_checks)
+@flashinfer_api
```
[the remark this answers is not in the document]
Comment (b8zhong) [reply flashinfer/gemm/routergemm.py]: Just added it

File: tests/model_optimizations/test_dsv3_router_gemm.py (hunk 2 above)
Comment (b8zhong) [reply tests/model_optimizations/test_dsv3_router_gemm.py]: 1. Renamed the file. For **unified API**, I did not add this. Could someone else help do this?

**Comment by dierksen** [comment]
@flashinfer-bot run
